File: internal/scheduler/retention.go
package scheduler

import (
	"encoding/json"
	"log"
	"sync"
	"time"
	"vaportrail/internal/db"

	"github.com/jonboulle/clockwork"
)

type RetentionManager struct {
	db    db.Store
	clock clockwork.Clock
	stop  chan struct{}
	wg    sync.WaitGroup
}

func NewRetentionManager(database db.Store) *RetentionManager {
	return &RetentionManager{
		db:    database,
		clock: clockwork.NewRealClock(),
		stop:  make(chan struct{}),
	}
}

func (rm *RetentionManager) Start() {
	rm.wg.Add(1)
	go rm.run()
}

func (rm *RetentionManager) Stop() {
	close(rm.stop)
	rm.wg.Wait()
}

func (rm *RetentionManager) run() {
	defer rm.wg.Done()
	// Run retention checks every hour
	ticker := rm.clock.NewTicker(1 * time.Hour)
	defer ticker.Stop()

	// Initial run
	rm.enforceRetention()

	for {
		select {
		case <-rm.stop:
			return
		case <-ticker.Chan():
			rm.enforceRetention()
		}
	}
}

func (rm *RetentionManager) enforceRetention() {
	targets, err := rm.db.GetTargets()
	if err != nil {
		log.Printf("RetentionManager: Failed to get targets: %v", err)
		return
	}

	for _, t := range targets {
		policies := defaultPolicies
		if t.RetentionPolicies != "" && t.RetentionPolicies != "[]" {
			var p []RetentionPolicy
			if err := json.Unmarshal([]byte(t.RetentionPolicies), &p); err == nil && len(p) > 0 {
				policies = p
			}
		}

		for _, p := range policies {
			cutoff := rm.clock.Now().Add(-time.Duration(p.Retention) * time.Second)

			if p.Window == 0 {
				// Raw data retention
				// We need a DeleteRawResults method in DB
				if err := rm.db.DeleteRawResultsBefore(t.ID, cutoff); err != nil {
					log.Printf("RetentionManager: Failed to delete raw results for %s: %v", t.Name, err)
				}
			} else {
				// Aggregated data retention
				// We need a DeleteAggregatedResultsBefore method
				if err := rm.db.DeleteAggregatedResultsBefore(t.ID, p.Window, cutoff); err != nil {
					log.Printf("RetentionManager: Failed to delete aggregated results (w=%d) for %s: %v", p.Window, t.Name, err)
				}
			}
		}
	}
}

File: vaportrail-rs/src/scheduler/retention.rs
//! Retention manager for cleaning up old data.

use crate::db::Store;

use super::rollup::get_retention_policies;
use chrono::{Duration as ChronoDuration, Utc};
use std::sync::Arc;
use std::time::Duration;
use tokio::sync::Mutex;

/// Manager for deleting data past retention periods.
pub struct RetentionManager {
    store: Arc<Store>,
    stop: Arc<Mutex<Option<tokio::sync::broadcast::Sender<()>>>>,
}

impl RetentionManager {
    pub fn new(store: Arc<Store>) -> Self {
        Self {
            store,
            stop: Arc::new(Mutex::new(None)),
        }
    }

    /// Start the retention manager background task.
    pub fn start(&self) {
        let store = self.store.clone();
        let stop = self.stop.clone();

        tokio::spawn(async move {
            let (tx, _) = tokio::sync::broadcast::channel(1);
            {
                let mut stop_guard = stop.lock().await;
                *stop_guard = Some(tx.clone());
            }

            let mut rx = tx.subscribe();
            let mut interval = tokio::time::interval(Duration::from_secs(60));

            loop {
                tokio::select! {
                    _ = rx.recv() => break,
                    _ = interval.tick() => {
                        process_retention(&store);
                    }
                }
            }
        });
    }

    /// Stop the retention manager.
    pub async fn stop(&self) {
        let stop = self.stop.lock().await;
        if let Some(tx) = stop.as_ref() {
            let _ = tx.send(());
        }
    }
}

fn process_retention(store: &Store) {
    let targets = match store.get_targets() {
        Ok(t) => t,
        Err(e) => {
            tracing::error!("RetentionManager: Failed to get targets: {}", e);
            return;
        }
    };

    let now = Utc::now();

    for target in targets {
        let policies = match get_retention_policies(&target) {
            Some(p) => p,
            None => continue,
        };

        for policy in policies {
            let cutoff = now - ChronoDuration::seconds(policy.retention as i64);

            if policy.window == 0 {
                // Delete raw results
                if let Err(e) = store.delete_raw_results_before(target.id, cutoff) {
                    tracing::error!(
                        "RetentionManager: Failed to delete raw results for {}: {}",
                        target.name,
                        e
                    );
                }
            } else {
                // Delete aggregated results for this window
                if let Err(e) = store.delete_aggregated_results_before(target.id, policy.window, cutoff) {
                    tracing::error!(
                        "RetentionManager: Failed to delete aggregated results for {} (w={}): {}",
                        target.name,
                        policy.window,
                        e
                    );
                }
            }
        }
    }
}
